Concept: functions (#715)
exercism/bash: concepts/functions/introduction.md

# Functions

Many Bash scripts are written in a strictly imperative style: execute one command, then execute another command, and so on.
But often you'll need to have a group of commands that conceptually perform a single purpose.
This is where _functions_ come in.

## Defining a Function

> You declare a function is one of two ways.

Reviewer: ```suggestion
You declare a function in one of two ways.
```

The first is a "portable" style

```bash

> funcname () { COMMANDS; }

Reviewer: Can we use `function_name` here?

It also shows what is,  I think, the style used for multiple names, while also using full words instead of abbreviations, so is helpful as documentation in its clarity as well.

Showing what I mean in a later suggestion...

Author: good point. Perhaps `my_function` since it doesn't _start_ with "function".

```

This is the style that was created with the original Bourne shell.

Alternately, you can use the `function` keyword

```bash

> function funcname { COMMANDS; }

Reviewer: ```suggestion
function function_name { COMMANDS; }
```

```

> There is no difference between the two styles.

Author: I don't think so.

> There is no difference between the two styles.

Reviewer 2: My personal preference, based on my time in the IRC channel, is for the portable version.

Mentioning two approaches without any suggestion in one direction or the other can leave someone new to the syntax a bit ... confused/adrift about which to use.

## Function Parameters

Functions, once defined, act like any other command (builtin or not).
Like any command, you can provide _arguments_ for your functions.
Inside the functions, you access the arguments using the _positional parameters_, `$1`, `$2`, etc.
(Recall, we learned about positional parameters in the [Variables][variables] concept.)

~~~~exercism/advanced
The special parameter `$0` is not changed inside a function; it is still the name of the executing script.
The currently executing function can access its name with the `$FUNCNAME` variable.

> ~~~~

Reviewer: We could give a link to the special parameters section manual page:

https://www.gnu.org/software/bash/manual/html_node/Special-Parameters.html

## Variables

You can define variables inside a function.
If you declare the variables with the `local` command then the _scope_ of the variable is limited to the current function (and to any functions called from it).
Otherwise, the variable is placed in the _global scope_.

Local variables can have the same name as a global variable.

> In that case, the local variable _temporarily_ overrides the global one, and the global value is restored when the function returns.

Reviewer: Having some thoughts about this line as well.

> In that case, the local variable _temporarily_ overrides the global one, and the global value is restored when the function returns.

Reviewer: Maybe it is covered well enough in lines +72 to +73

```bash
x=5

myfunc () {
    local x=100
    echo "in my function, $x == 100"
}

echo "in the global scope, $x == 5"

myfunc

echo "back in the global scope, $x == 5"
```

Inside a function, you can access variables from the _caller_'s scope.
That means you can use global variables, or local variables that are declared in some function that calls this one.

~~~~exercism/advanced
Technically, "global" is not the right word to use.
Assignments to non-local variables will assign to the variable with that name that has been declared in some previous scope, up to the global scope.

> This example is taken from the bash manual

Reviewer: Which section of the man page?  I am not finding this in the man pages section 1.  Or a better question which bash manual?

Can we give the command or link to get to it?

```bash
func1() {
    local var='func1 local'
    func2
}

func2() {
    echo "In func2, var = $var"
}

var=global
func1
func2
```

The output is:

```none
In func2, var = func1 local
In func2, var = global
```
~~~~

> ## Return Value

Author: a function can have multiple parameters and multiple local variables but only one return value.

> ## Return Value

Reviewer 2: A function call has one return value. A function can have many calls and many return values.

A function variable has one name and one value (at any given time).

It feels very odd to me for them to be inconsistent.

A function, like any command, has an _exit status_.
By default, the status of a function is the exit status of the _last command executed_.

You can use the `return` command to return from a function with a specific exit status.

```bash
check_password () {
    if [[ $1 == "secret" ]]; then
        return 0
    else
        return 1
    fi
}

read -sp "Enter your password: " pass

if check_password "$pass"; then
    echo "Correct!"
else
    echo "Wrong password."
fi
```

~~~~exercism/note
Note that the `check_password` function can be simplified to:

```bash
check_password () { [[ $1 == "secret" ]]; }
```

1. The `[[...]]` conditional construct has an exit status: `0` for "true", `1` for "false.
2. The `{...}` grouping construct must have either a newline or a semicolon before the ending brace.
~~~~

## Function Output

The return status of a function is just a number.
How can a function produce output?

> Your function simply emits output on standard output.

Reviewer 2: ```suggestion
Your function can print to standard output.
Then, you can use command substitution to capture it. 
```

Use the familiar _command substitution_ to capture it:

```bash

> d6 () { echo $(( 1 + RANDOM % 6 )); }

Reviewer 2: Always quote every expansion ;)

die=$( d6 )
echo "You rolled a $die."
```

### Using Both the Output and the Status

> The exit status of a function is still available to use when you are capturing the output.

Reviewer 2: ```suggestion
The exit status of a function is available to use even when you are capturing the output.
```

```bash
roll () {
    local n=$1
    if (( 4 <= n && n <= 20 )); then
        echo $(( 1 + RANDOM % n ))  # exit status is 0
    else
        return 1
    fi
}

read -p "How many faces does your die have? " faces
if die=$( roll "$faces" ); then
    echo "You rolled a $die."
else
    echo "I can't roll a die with $faces faces."
fi
```

> ## Recursion

Reviewer 2: No examples? Fibonacci would be a classic.

Functions can call themselves recursively.
By default, there is no limit to the depth of recursion.

> 

Reviewer: ```suggestion
```

To expand a variable in a function, Bash will traverse up the call stack, as far as the global scope, to find a function where that variable name has been declared.

This example is adapted from the [Shell Functions][man-funcs] section of the manual:

```bash
func1() {
    local var='func1 local'
    func2
}

func2() {
    echo "In func2, var = $var"
}

var=global
func1
func2
```

The output is:

```none
In func2, var = func1 local
In func2, var = global
```

Similarly, _assigning_ a value to a variable will assign it _in the scope where it was declared_.
This "action at a distance" can create hard-to-follow code, as it is not always obvious where a variable was assigned a value.
~~~~

## Return Value

A function, like any command, has an _exit status_.
By default, the status of a function is the exit status of the _last command executed_.

You can use the `return` command to return from a function with a specific exit status.

```bash
check_password () {
    if [[ $1 == "secret" ]]; then
        return 0
    else
        return 1
    fi
}

read -sp "Enter your password: " pass

if check_password "$pass"; then
    echo "Correct!"
else
    echo "Wrong password."
fi
```

> Using `return` with no arguments returns a zero status.

Reviewer 2: ```suggestion
Using `return` with no arguments returns the status of the last command executed.
```

~~~~exercism/note
Note that the `check_password` function can be simplified to:

```bash
check_password () { [[ $1 == "secret" ]]; }
```

1. The `[[...]]` conditional construct has an exit status: `0` for "true", `1` for "false.
2. The `{...}` grouping construct must have either a newline or a semicolon before the ending brace.
~~~~

## Function Output

The return status of a function is just a number.
How can a function produce output?

Your function can print to standard output.
Use the familiar _command substitution_ to capture it:

```bash
d6 () { echo "$(( 1 + RANDOM % 6 ))"; }

die=$( d6 )
echo "You rolled a $die."
```

### Using Both the Output and the Status

The exit status of a function is available to use even when you are capturing the output.

```bash
roll () {
    local n=$1
    if (( 4 <= n && n <= 20 )); then
        echo "$(( 1 + RANDOM % n ))"  # exit status is 0
    else
        return 1
    fi
}

read -p "How many faces does your die have? " faces
if die=$( roll "$faces" ); then
    echo "You rolled a $die."
else
    echo "I can't roll a die with $faces faces."
fi
```

## Recursion

Functions can call themselves recursively.

> By default, there is no limit to the depth of recursion.

Reviewer 2: No recursion limit ... but factorials would overflow pretty quickly :smile:

An example:

```bash
factorial() {
    local n=$1

>     if ((n <= 1)); then

Reviewer 2: The prior snippet using `((` space pads the `((` command.

```suggestion
    if (( n <= 1 )); then
```

        echo "$(( 1 + RANDOM % n ))"  # exit status is 0
    else
        return 1
    fi
}

read -p "How many faces does your die have? " faces
if die=$( roll "$faces" ); then
    echo "You rolled a $die."
else
    echo "I can't roll a die with $faces faces."
fi
```

## Recursion

Functions can call themselves recursively.
By default, there is no limit to the depth of recursion.

An example:

```bash
fibonacci() {
    local n=$1
    if (( n <= 1 )); then
        echo "1"
    else
        local a=$(fibonacci "$(( n - 1 ))")
        local b=$(fibonacci "$(( n - 2 ))")
        echo "$(( a + b ))"
    fi
}

for i in {1..10}; do fibonacci "$i"; done
# => 1
# => 2
# => 3
# => 5
# => 8
# => 13
# => 21
# => 34
# => 55
# => 89
```

~~~~exercism/advanced
The recursion depth can be controlled with [the `FUNCNEST` variable][funcnest].

```bash
bash -c '
    recur() {
        echo $1
        recur $(($1 + 1))
    }
    FUNCNEST=5
    recur 1
'
```

```none
1
2
3
4
5
environment: line 1: recur: maximum function nesting level exceeded (5)
```

[funcnest]: https://www.gnu.org/software/bash/manual/bash.html#index-FUNCNEST
~~~~

[variables]: https://exercism.org/tracks/bash/concepts/variables

> [man-funcs]: https://www.gnu.org/software/bash/manual/bash.html#Shell-Functions

Reviewer: Same here.

Otherwise, the variable is placed in the _global scope_.

Local variables can have the same name as a global variable.

> In that case, the local variable takes precedence over the global variable.

Reviewer: Does it first use the local variable and then use the global variable?  To precede something means to come first, implying that the latter comes as well.  I think that was why the wording was used to not imply that something comes first, and that the global is not available at that scope.

At least that is how I understood it, and was personally glad to not see "precedence" used here. I can not speak for the author, though.

Reviewer: Is it perhaps "not available" or "not visible" due to the scope?

Author: This is a hard concept to explain. The latest commit copies sentences directly from the manual.

Reviewer: I agree with that.

I sometimes do so graphically, and the shadowing idea follows that.
